### PR TITLE
Fix running bootstrap tests with a local Rust toolchain as the stage0

### LINKED ISSUE #105766: Some bootstrap tests fail with custom rustc
### Command

I tried this with Rust 1.66.0, simplified from my build for Fedora:

```shell
$ ./configure --local-rust-root=/usr
$ ./x test bootstrap
```

I expected to see this happen: passing tests

Instead, this happened: 20 `builder::tests::...` failures, all of this sort:

```
---- builder::tests::alias_and_path_for_library stdout ----
failed to execute command: "/builddir/build/BUILD/rustc-1.66.0-src/build/x86_64-unknown-linux-gnu/stage0/bin/rustc" "--target" "x86_64-unknown-linux-gnu" "--print" "target-libdir"
error: No such file or directory (os error 2)
thread 'main' panicked at 'status code: 1', lib.rs:1654:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

My stage0 should be using `/usr/bin/rustc`.

### Version it worked on

It most recently worked on: Rust 1.65.0

### Version with regression

`rustc-1.66.0-src/version`:
```
1.66.0 (69f9c33d7 2022-12-12)
```

I bisected this to #98483, commit 55c040e52964e4ad81b4b2af3c91068a80b92f11, which stopped using `CARGO` and `RUSTC` environment variables when building a `Config`.

### Backtrace
<details><summary>Backtrace</summary>
<p>

```
thread 'main' panicked at 'status code: 1', lib.rs:1654:9
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: bootstrap::detail_exit
             at ./lib.rs:1654:9
   3: bootstrap::util::fail
             at ./util.rs:475:5
   4: bootstrap::util::output
             at ./util.rs:425:19
   5: bootstrap::Build::new
             at ./lib.rs:436:13
   6: bootstrap::builder::tests::configure_with_args
             at ./builder/tests.rs:18:27
   7: bootstrap::builder::tests::configure
             at ./builder/tests.rs:6:5
   8: bootstrap::builder::tests::alias_and_path_for_library
             at ./builder/tests.rs:142:55
   9: bootstrap::builder::tests::alias_and_path_for_library::{{closure}}
             at ./builder/tests.rs:140:1
  10: core::ops::function::FnOnce::call_once
             at /builddir/build/BUILD/rustc-1.65.0-src/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

</p>
</details>

@rustbot modify labels: +regression-from-stable-to-stable -regression-untriaged +A-bootstrap

### PR DESCRIPTION
**Pull request by emilyalbini**
When configuring a local Rust toolchain as the stage0 (with `build.rustc` and `build.cargo` in `config.toml`) we noticed there were test failures (both on the Python and the Rust side) due to bootstrap not being able to find rustc and Cargo.

This was due to those two `config.toml` settings not being propagated in the tests. This PR fixes the issue by ensuring rustc and cargo are always configured in tests, using the parent bootstrap's `initial_rustc` and `initial_cargo`.

try-job: x86_64-msvc
Fixes https://github.com/rust-lang/rust/issues/105766

### CODE REVIEW

**Comment by cuviper** [comment]
I suppose this will fix #105766?

**Comment by matthiaskrgr** [comment]
@bors r- I guess
https://github.com/rust-lang/rust/pull/126486#issuecomment-2168952828

**Comment by onur-ozkan** [comment]
Seems like it's missing `.exe` part.

**Comment by pietroalbini** [comment]
> I suppose this will fix https://github.com/rust-lang/rust/issues/105766?

Yes!

**Comment by pietroalbini** [comment]
@rustbot author

Figured out another case that'd need tweaking.

**Comment by pietroalbini** [comment]
Confirmed this fixes all of the problems with using custom rustc/cargo in an airgapped environment.

@rustbot review

File: src/bootstrap/src/core/config/config.rs [hunk 1]
```diff
@@ -1440,14 +1451,24 @@ impl Config {
             config.out = crate::utils::helpers::absolute(&config.out);
         }
 
+        // Hacky way to determine the executable suffix for the build target. We cannot use
+        // std::env::consts::EXE_SUFFIX as the build target might not be the target bootstrap was
+        // compiled with.
+        let initial_exe_suffix = if config.build.triple.contains("windows") { ".exe" } else { "" };
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/config/config.rs:1457]: You can use `utils::helpers::exe` function.

ref:
https://github.com/rust-lang/rust/blob/b4b2643c111714b541db1c18c178ad0ce654238b/src/bootstrap/src/core/config/config.rs#L1941

**Review by onur-ozkan — CHANGES_REQUESTED** [review]

File: src/bootstrap/src/core/config/config.rs (hunk 1 above)
Comment (pietroalbini) [reply src/bootstrap/src/core/config/config.rs]: TIL about that helper! Updated.

**Comment by onur-ozkan** [comment]
LGTM, thanks!

@bors r+

**Pull request merged**